### ISSUE TITLE
Weird behaviour when bash function has the Unicode name

### ISSUE THREAD

**Issue by giappi** [koalaman/shellcheck#2426]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1036, SC1088
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

function ■()
{
    echo "■ $@"
}

```

#### Here's what shellcheck currently says:
```
Line 3	SC1036: '(' is invalid here. Did you forget to escape it?
Line 3	SC1088: Parsing stopped here. Invalid use of parentheses?
```
#### Here's what I wanted or expected to see:

Unicode is valid bash function name. In case of it does not support Unicode, I think it should show the error message like "invalid function name".

**Comment by grondilu** [comment]
I'm not too surprised shellcheck does not accept that.  It should, though, as in the bash man page, one can read:

> In default mode, a function name can be any unquoted shell word  that  does not  contain $

I could not quite find what a shell word is, as the definition is a bit vague:

> A sequence of characters considered as  a  single  unit  by  the shell.

I suspect it is pretty much a any sequence of unicode characters.

This is confirmed if you try very weird function names:

```bash
[]{}#()  { echo "ok"; }
```

This actually defines a function `[]{}#`, amazingly enough.

Needless to say, shellcheck refuses to parse that.

**Comment by GB609** [comment]
I'm running into this as well, as i'm trying to name functions with '#' in it, e.g.
```bash
function prefix#functionName { 
   echo TBD: 
}
```
This is a perfectly valid function name for bash in non-POSIX mode, because '#' is ony marking a comment when:
- at the start of a line
- preceeded by whitespace
- preceeded by a special operator, which are several different tokens containing `&`, `;`, `|`, `(`, `)` in unescaped/unquoted form (plus redirection statements like `>&2`).

In all other situations, e.g. when preceeded by any ascii letter, '#' is treated as just another character.

The snipped above leads to 
```
^-- SC1073 (error): Couldn't parse this function. Fix to allow more checks.
```

One can trick shellcheck into continuing to parse the file by a little reformatting:
```bash
function prefix#functionName 
{ 
   echo TBD: 
}
prefix#functionName
```
But this messes up the invocation check, since the function is just named `prefix` for shellcheck.
So, i get SC2329 when using `prefix#functionName`:
```
^-- SC2329 (info): This function is never invoked. Check usage (or ignored if invoked indirectly).
```

using just `prefix` as invocation is detected then for `prefix#functionName`.

Adding a bash-shebang to the file doesn't change anything.

**Comment by hashar** [comment]
Here is an unicode related one with SC2034, shellcheck can't find the function `λ` has been called:
```bash
#!/bin/bash

# SC2034 (warning): host_id appears unused.
#                   Verify use (or export if used externally).
λ() {
	host_id=$(id)
}

λ
```
